initramfs/src/filesystem.rs

```diff
@@ -0,0 +1,68 @@
+use bytes::Bytes;
+use cpio::NewcBuilder;
+use flate2::read::GzDecoder;
+use libflate::gzip;
+use log::debug;
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{Cursor, Read},
+};
+use tar::Archive;
+
+pub struct Filesystem {
+    files: HashMap<String, (NewcBuilder, Cursor<Vec<u8>>)>,
+}
+
+impl Filesystem {
+    pub fn new() -> Self {
+        Self {
+            files: HashMap::new(),
+        }
+    }
+
+    pub fn parse_files(&mut self, archives: Vec<Archive<GzDecoder<Cursor<Bytes>>>>) {
+        // We need to reverse the layers so that the last layer is the first one in the cpio
+        for mut archive in archives.into_iter().rev() {
+            // Iterate over all the files in the layer
+            for entry in archive.entries().unwrap() {
+                let mut entry = entry.unwrap();
+                let path = entry.header().path().unwrap().to_str().unwrap().to_string();
+
+                // If the file already exists, we don't need to add it again
+                if self.files.contains_key(&path) {
+                    continue;
+                }
+
+                // Create a new cpio builder for the file
+                let builder = NewcBuilder::new(entry.header().path().unwrap().to_str().unwrap())
+                    .uid(entry.header().uid().unwrap() as u32)
+                    .gid(entry.header().gid().unwrap() as u32)
+                    .mode(entry.header().mode().unwrap());
+
+                // Read the file into a buffer
+                let mut contents = Vec::new();
+                entry.read_to_end(&mut contents).unwrap();
+
+                // Add the file to the cpio
+                self.files.insert(path, (builder, Cursor::new(contents)));
+            }
+        }
+
+        debug!("Successfully parsed all layers");
+    }
+
+    pub fn write_cpio(&mut self, file_name: &str) {
+        // Write the cpio to disk
+        let archive = File::create(file_name).unwrap();
```

Comment:
idem, you should prevent using unwrap and handle errors gracefully

Reply:
I fixed this while force pushing for the last time